### ISSUE TITLE
[java] ShortVariable should whitelist lambdas

### ISSUE THREAD

**Issue by jsotuyod** [pmd/pmd#720]
**Rule Set:**
java-naming

**Description:**
Lambdas can be short snippets using short variable names for known parameter types (such as calling a `String` `s`, or an `int` `i`. The rule currently ignores for-loops, but doesn't consider lambdas in the same way.

**Code Sample demonstrating the issue:**

```java
aListFullOfStrings.forEach(s -> s.replaceAll(" ", ""));
```

Possible fix (to be thoroughly tested):

```xml
[not(ancestor::ForInit)]
**[not(ancestor::LambdaExpression)]** [not(../../VariableDeclarator and ../../../LocalVariableDeclaration and ../../../../ForStatement)]
[not((ancestor::FormalParameter) and (ancestor::TryStatement))]
```

**Comment by nachimehta** [comment]
Any official work arounds here? I find that if I create a ruleset.xml file and add

    <rule ref="rulesets/java/naming.xml/ShortMethodName">
        <properties>
            <property name="violationSuppressXPath" value=".[ancestor::LambdaExpression]"/>
        </properties>
    </rule>

That seems to work, but it also removes a check for shortVariableNames for some reason. i.e.

    String s = "helloworld";

Doesn't get caught anymore. I've tried a variety of configurations to no avail.

**Comment by nachimehta** [comment]
Turns out the error is coming from VariableName rule:

    <rule ref="rulesets/java/naming.xml/ShortVariable">
        <properties>
            <property
                    name="violationSuppressXPath" value=".[ancestor::LambdaExpression]"/>
        </properties>
    </rule>

Works beautifully

**Comment by oowekyala** [comment]
`[not(ancestor::LambdaExpression)]` ignores all variables declared in a lambda. Maybe we should restrict that to lambda parameters, eg `[not(../../LambdaExpression/FormalParameter)]` ? If the lambda is a block lambda, then usual naming conventions should probably apply for variable declarations
